### ISSUE TITLE
ScaleData with vars throws an error after Integration 

### ISSUE THREAD

**Issue by TapsiS** [satijalab/seurat#1507]
I am running the code which gives an error - 
```
tum.integrated1.1 <- ScaleData(tum.integrated, vars.to.regress  = "nFeature_RNA")
```
" Regressing out nFeature_RNA
  |                                                                      |   0%Error in qr.resid(qr = qr, y = data.expr[x, ]) :
  'qr' and 'y' must have the same number of rows

But when I don't give any vars to regress, it runs smoothly - 

```
tum.integrated1 <- ScaleData(tum.integrated1)
```
Centering and scaling data matrix
  |======================================================================| 100%

My dataset has 27 samples and 75k cells and no NA's.

**Comment by satijalab** [comment]
I cannot reproduce this error on my end, for example

`immune.combined <- ScaleData(immune.combined,vars.to.regress = 'nFeature_RNA')` runs fine, using the output of our ctrl/stim PBMC tutorial.

I can suggest three things

1. Could you confirm the DefaultAssay of the object you are trying to run? 
2. If you are running on integrated data (as I assume you are), can you confirm there are no NAs in the `tum.integrated1[["integrated"]]@data` matrix,
3. Can you check to make sure that all values of `tum.integrated1$nFeature_RNA` are defined?
4. Does this error pop up regardless of what feature you choose to regress on?

**Comment by lb15** [comment]
Hello,

I'm getting the same error but through a different mechanism. I have two seurat objects that I'd like merge and then regress out batch effects. I used SCTransform on both objects individually and then did the following:

```
seur = merge(ob1, ob2)
seur <- SCTransform(seur, vars.to.regress = "percent.mt")
```

I've tried to regress out different features, including percent.mt and orig.ident but they both result in same error. I've tried with the default assay set as either "SCT" or "RNA", but both led to same error.

Thanks for any advice!

**Comment by TapsiS** [comment]
The error only occurs on the integrated assay with vars to regress. 

I checked the assay - 
1. DefaultAssay is integrated
2. There are no NA's
3. It is defined since I tried with another variable such as sample_id and I still get the same error.
I have attached the traceback, with a sample_id variable.

traceback()
16: stop(condition)
15: resignalConditions(future)
14: value.Future(tmp, ...)
13: value(tmp, ...)
12: values.list(fs)
11: values(fs)
10: future_lapply(X = 1:ncol(x = chunk.points), FUN = function(i) {
        return(RegressOutMatrix(data.expr = object[chunk.points[1,
            i]:chunk.points[2, i], , drop = FALSE], latent.data = latent.data,
            model.use = model.use, use.umi = use.umi, verbose = FALSE))
    })
9: ScaleData.default(object = GetAssayData(object = object, slot = slot.use),
       features = features, vars.to.regress = vars.to.regress, latent.data = latent.data,
       model.use = model.use, use.umi = use.umi, do.scale = do.scale,
       do.center = do.center, scale.max = scale.max, block.size = block.size,
       min.cells.to.block = min.cells.to.block, verbose = verbose,
       ...)
8: ScaleData(object = GetAssayData(object = object, slot = slot.use),
       features = features, vars.to.regress = vars.to.regress, latent.data = latent.data,
       model.use = model.use, use.umi = use.umi, do.scale = do.scale,
       do.center = do.center, scale.max = scale.max, block.size = block.size,
       min.cells.to.block = min.cells.to.block, verbose = verbose,
       ...)
7: IsMatrixEmpty(x = new.data)
6: SetAssayData.Assay(object = object, slot = "scale.data", new.data = ScaleData(object = GetAssayData(object = object,
       slot = slot.use), features = features, vars.to.regress = vars.to.regress,
       latent.data = latent.data, model.use = model.use, use.umi = use.umi,
       do.scale = do.scale, do.center = do.center, scale.max = scale.max,
       block.size = block.size, min.cells.to.block = min.cells.to.block,
       verbose = verbose, ...))
5: SetAssayData(object = object, slot = "scale.data", new.data = ScaleData(object = GetAssayData(object = object,
       slot = slot.use), features = features, vars.to.regress = vars.to.regress,
       latent.data = latent.data, model.use = model.use, use.umi = use.umi,
       do.scale = do.scale, do.center = do.center, scale.max = scale.max,
       block.size = block.size, min.cells.to.block = min.cells.to.block,
       verbose = verbose, ...))
4: ScaleData.Assay(object = assay.data, features = features, vars.to.regress = vars.to.regress,
       latent.data = latent.data, model.use = model.use, use.umi = use.umi,
       do.scale = do.scale, do.center = do.center, scale.max = scale.max,
       block.size = block.size, min.cells.to.block = min.cells.to.block,
       verbose = verbose, ...)
3: ScaleData(object = assay.data, features = features, vars.to.regress = vars.to.regress,
       latent.data = latent.data, model.use = model.use, use.umi = use.umi,
       do.scale = do.scale, do.center = do.center, scale.max = scale.max,
       block.size = block.size, min.cells.to.block = min.cells.to.block,
       verbose = verbose, ...)
2: ScaleData.Seurat(tum.integrated, vars.to.regress = "orig.ident",
       assay = "integrated")
1: ScaleData(tum.integrated, vars.to.regress = "orig.ident", assay = "integrated")

**Comment by satijalab** [comment]
Apologies - I can't think of a good explanation here. I can only recommend the following debugging procedure: presumably there is some individual feature that is causing an error for regression (perhaps it is 0 across all cells or something similar). I would try seeing if you could reduce your feature set (perhaps run on  features 1-100, 101-200, etc.), and see if you can narrow down which bin is throwing the error. Then explore further within each bin, to find the exact feature causing the error.

**Comment by satijalab** [comment]
I am closing this issue for now, as we have not encountered it in any other context. Feel free to get in touch if it continues to be a problem, if you are able to send us the object we can repeat the debugging procedure above

**Comment by yulongliu68** [comment]
Hi all, I was having the same issue. 
`zx24_40ov <- merge(x=zx2_40ov,y=zx4_40ov)`
`zx24_40ov <- PercentageFeatureSet(zx24_40ov, pattern = "^mt-", col.name = "percent.mt")`
`zx24_40ov <- SCTransform(zx24_40ov,  verbose = FALSE, vars.to.regress = "percent.mt")`

> Error in qr.resid(qr = qr, y = data.expr[x, ]) : 
  'qr' and 'y' must have the same number of rows

I found a workaround that solves the problem by just remaking the Seurat object from the count data. 

`zx24_40ov_new <- CreateSeuratObject(zx24_40ov@assays$RNA@counts, project = "zx24_40ov")`
`zx24_40ov_new$orig.ident <- zx24_40ov$orig.ident`

**Comment by dagnyr** [comment]
I just had this exact issue today! Just wanted to note that I think some of my metadata columns may have been contributing. When I followed @yulongliu68's solution above and copied all of my metadata, the issue persisted, but when I only copied a few select columns, I had no issue. For the one's causing the issue, I just manually added in the information I wanted for each sample using some loops. Hope this helps anyone experiencing this issue!
